src/VisualStudio/Core/Next/Remote/JsonRpcClient.cs

```diff
@@ -88,23 +93,71 @@ public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments,
             return Extensions.InvokeAsync(_rpc, targetName, arguments, funcWithDirectStreamAsync, _cancellationToken);
         }
 
+        public void Dispose()
+        {
+            OnDisposed();
+
+            _rpc.Dispose();
+        }
+
+        protected void StartListening()
+        {
+            // due to this issue - https://github.com/dotnet/roslyn/issues/16900#issuecomment-277378950
+            // _rpc need to be explicitly started
+            _rpc.StartListening();
+        }
+
+        protected virtual void OnDisposed()
+        {
+            // do nothing
+        }
+
+        // these are for debugging purpose. once we find out root cause of the issue
+        // we will remove these.
+        private static JsonRpcDisconnectedEventArgs s_debuggingLastDisconnectReason;
+        private static string s_debuggingLastDisconnectCallstack;
+
+        private JsonRpcDisconnectedEventArgs _debuggingLastDisconnectReason;
+        private string _debuggingLastDisconnectCallstack;
+
         private bool ReportUnlessCanceled(Exception ex, CancellationToken cancellationToken)
         {
             if (cancellationToken.IsCancellationRequested)
             {
                 return true;
             }
 
-            // save extra info using NFW
-            ReportExtraInfoAsNFW(ex);
+            s_debuggingLastDisconnectReason = _debuggingLastDisconnectReason;
+            s_debuggingLastDisconnectCallstack = _debuggingLastDisconnectCallstack;
 
-            // make it to explicitly crash to get better info
-            FatalError.Report(ex);
+            // send NFW to figure out why this is happening
+            ReportExtraInfoAsNFW(ex);
 
             GC.KeepAlive(_debuggingLastDisconnectReason);
             GC.KeepAlive(_debuggingLastDisconnectCallstack);
 
-            return Contract.FailWithReturn<bool>("shouldn't be able to reach here");
+            return true;
+        }
+
+        private static bool s_reported = false;
+
+        private void ThrowOwnCancellationToken()
+        {
+            if (CodeAnalysis.PrimaryWorkspace.Workspace != null && !s_reported)
+            {
+                // do not report it multiple times
+                s_reported = true;
+
+                // use info bar to show warning to users
+                CodeAnalysis.PrimaryWorkspace.Workspace.Services.GetService<IErrorReportingService>()?.ShowGlobalErrorInfo(
+                    ServicesVSResources.Unfortunately_a_process_used_by_Visual_Studio_has_encountered_an_unrecoverable_error_We_recommend_saving_your_work_and_then_closing_and_restarting_Visual_Studio);
```

Comment:
Is this really any better than what's currently happening?  Showing the goldbar and NFW'ing is fine to me if VS is actually recoverable.  But it sounds like it isn't and the user needs to restart it anyways.